## **Mô tả bài toán**
Một công ty bán hàng online tại UK cần đánh giá tính hình kinh doanh của họ, bộ dữ liệu được đưa cho người phân tích gồm các thông tin sau:

- InvoiceNo: Số hóa đơn
- StockCode: mã hàng
- Description: Mô tả hàng
- Quantity: Số lương
- InvoiceDate: Ngày bán
- UnitPrice: Đơn giá
- CustomerID: Mã khách
- Country: Nước sản xuất

Bạn hãy thực hiện các công việc sau:
1. Đọc dữ liệu từ file OnlineRetail.csv
2. Tìm hiểu cấu trúc bộ dữ liệu: Số lượng quan sát, số thuộc tính trong bộ dữ liệu, thang đo tương ứng của các thuộc tính
3. Hãy giải đáp các thắc mắc sau:Công ty bán hàng do bao nhiêu nước sản xuất
Tổng số lượng đơn hàng bán ra, tổng doanh thu
Top 10 mặt hàng có số lượng bán ra lớn nhất
Top 10 mặt hàng có doanh thu lớn nhất'''

In [20]:
'''1. Đọc dữ liệu từ file OnlineRetail.csv'''
import pandas as pd
data=pd.read_csv('OnlineRetail.csv', encoding = "ISO-8859-1")

In [4]:
# hiển thị 5 dòng dữ liệu đầu tiên
data.head(5)

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [21]:
'''2. Tìm hiểu cấu trúc bộ dữ liệu'''
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


- Bộ dữ liệu chứa 541909 dòng dữ liệu và 8 cột, chỉ có cột CustomerID không có đầy đủ thông tin (406829 non-null <541909 )
- InvoiceNo, StockCode, Description, InvoiceDate, country: là thuộc tính định tính, có thang đo định danh
- CustomerID mặc dù có kiểu số (float64), nhưng các con số này chỉ có ý nghĩa phân biệt khác hàng này với khách hàng khác –> thuộc tính định danh
- Quantity, UnitPrice: thuộc tính định lượng, thang đo ratio

### **Công ty bán hàng do bao nhiêu nước sản xuất**

In [22]:
'''3. Giải đáp thắc mắc'''
# 3.1 Công ty bán hàng tại bao nhiêu quốc gia
# Lấy ra tên các quốc gia
country = data.Country.unique()
print ("số lượng các quốc gia: " + str(country.size))

số lượng các quốc gia: 38


### **Tổng số lượng đơn hàng bán ra, tổng doanh thu**

In [23]:
# 3.2 Số lượng đơn hàng bán ra và tổng doanh thu
# Tạo cột tính thành tiền của các mặt hàng
data['Total'] = data['Quantity'] * data['UnitPrice']
data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  Total  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  15.30  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  22.00  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom  20.34

In [45]:
data.InvoiceDate.value_counts()

10/31/2011 14:41    1114
12/8/2011 9:28       749
12/9/2011 10:03      731
12/5/2011 17:24      721
6/29/2011 15:58      705
                    ... 
9/28/2011 9:57         1
9/28/2011 9:45         1
3/16/2011 12:13        1
3/16/2011 12:22        1
3/16/2011 15:58        1
Name: InvoiceDate, Length: 23260, dtype: int64

In [24]:
# Giá trị đơn hàng của mỗi đơn hàng
total_invoices = data['Total'].sum()
print ("Số lượng hóa đơn bán ra: "+ str (total_invoices.size))
print ("Tổng doanh thu: " + str(total_invoices.sum()))

Số lượng hóa đơn bán ra: 1
Tổng doanh thu: 9747747.933999998


### **Top 10 mặt hàng có số lượng hóa đơn bán ra nhiều nhất**

In [25]:
# 3.3 Method 1: Top 10 mặt hàng có số lượng bán ra lớn nhất
quantity_product = data.sort_values('Quantity',ascending= False)
quantity_product[['StockCode', 'Description', 'Quantity']].head(10)

StockCode                          Description  Quantity
540421     23843          PAPER CRAFT , LITTLE BIRDIE     80995
61619      23166       MEDIUM CERAMIC TOP STORAGE JAR     74215
502122     84826       ASSTD DESIGN 3D PAPER STICKERS     12540
74614      37413                                  NaN      5568
421632     84077    WORLD WAR 2 GLIDERS ASSTD DESIGNS      4800
206121     22197                 SMALL POPCORN HOLDER      4300
220843    85123A                                    ?      4000
97432      22053                EMPIRE DESIGN ROSETTE      3906
270885     18007  ESSENTIAL BALM 3.5g TIN IN ENVELOPE      3186
160546     21108   FAIRY CAKE FLANNEL ASSORTED COLOUR      3114

In [41]:
# Nếu groupby 'StockCode' và 'Description' 
# thì kết quả trả về sẽ tự loại bỏ kết quả Description có chứa 'NaN' ? không đúng lắm
quantity_product2 = data.groupby(['StockCode', 'Description'])['Quantity'].max().sort_values(ascending= False)
quantity_product2.head(10)

StockCode  Description                        
23843      PAPER CRAFT , LITTLE BIRDIE            80995
23166      MEDIUM CERAMIC TOP STORAGE JAR         74215
84826      ASSTD DESIGN 3D PAPER STICKERS         12540
84077      WORLD WAR 2 GLIDERS ASSTD DESIGNS       4800
22197      SMALL POPCORN HOLDER                    4300
85123A     ?                                       4000
22053      EMPIRE DESIGN ROSETTE                   3906
18007      ESSENTIAL BALM 3.5g TIN IN ENVELOPE     3186
21108      FAIRY CAKE FLANNEL ASSORTED COLOUR      3114
23343      came coded as 20713                     3100
Name: Quantity, dtype: int64

### **Top 10 mặt hàng có tổng số lượng hóa đơn bán ra lớn nhất**

In [51]:
# 3.3 Method 2: Top 10 mặt hàng có tổng số lượng bán ra lớn nhất
quantity_product2 = data.groupby('StockCode')['Quantity'].sum().sort_values(ascending= False)
quantity_product2.head(10)

StockCode
22197     56450
84077     53847
85099B    47363
85123A    38830
84879     36221
21212     36039
23084     30646
22492     26437
22616     26315
21977     24753
Name: Quantity, dtype: int64

In [53]:
# 3.3 Method 2: Top 10 mặt hàng có tổng số lượng bán ra lớn nhất
quantity_product2 = data.groupby(['StockCode', 'Description'])['Quantity'].sum().sort_values(ascending= False)
quantity_product2.head(10)

StockCode  Description                       
84077      WORLD WAR 2 GLIDERS ASSTD DESIGNS     53847
85099B     JUMBO BAG RED RETROSPOT               47363
84879      ASSORTED COLOUR BIRD ORNAMENT         36381
22197      POPCORN HOLDER                        36334
21212      PACK OF 72 RETROSPOT CAKE CASES       36039
85123A     WHITE HANGING HEART T-LIGHT HOLDER    35025
23084      RABBIT NIGHT LIGHT                    30680
22492      MINI PAINT SET VINTAGE                26437
22616      PACK OF 12 LONDON TISSUES             26315
21977      PACK OF 60 PINK PAISLEY CAKE CASES    24753
Name: Quantity, dtype: int64

### **Top 10 mặt hàng có doanh thu lớn nhất**

In [54]:
# 3.3 Top 10 mặt hàng có doanh thu lớn nhất
# sử dụng groupby để tiến hành tính tổng (sum) theo nhóm.
#Vì mỗi mã hàng sẽ có một mô tả tương ứng, để rõ ràng, tiến hành gom nhóm theo 2 thuộc tính: StockCode, Description
total_quantity_product = data.groupby(['StockCode', 'Description'])['Total'].sum().sort_values(ascending= False)
total_quantity_product.head(10)

StockCode  Description                       
DOT        DOTCOM POSTAGE                        206245.48
22423      REGENCY CAKESTAND 3 TIER              164762.19
47566      PARTY BUNTING                          98302.98
85123A     WHITE HANGING HEART T-LIGHT HOLDER     97715.99
85099B     JUMBO BAG RED RETROSPOT                92356.03
23084      RABBIT NIGHT LIGHT                     66756.59
POST       POSTAGE                                66230.64
22086      PAPER CHAIN KIT 50'S CHRISTMAS         63791.94
84879      ASSORTED COLOUR BIRD ORNAMENT          58959.73
79321      CHILLI LIGHTS                          53768.06
Name: Total, dtype: float64

**Kết luận:**
- Chúng ta nhìn thấy rằng, mặt hàng bán ra với số lượng nhiều nhất, không phải mặt hàng mang lại doanh thu lớn nhất
- Có một số mặt hàng như 85099B – JUMBO BAG RED RETROSPOT và 23084 RABBIT NIGHT LIGHT nằm trong top 10 cả về số lượng lẫn doanh thu

In [55]:
pwd

'C:\\Users\\vai1hc\\Module 2'